polygeist/tools/cgeist/Lib/Attributes.h

```diff
@@ -9,90 +9,216 @@
+  /// Create a NamedAttribute with name \p AttrName and value \p Attr and add it
   /// to the builder.
-  AttrBuilder &addAttribute(llvm::Twine attrName, mlir::Attribute attr);
-
-  /// Add the given named attribute \p attr to the builder.
-  AttrBuilder &addAttribute(mlir::NamedAttribute attr) {
-    attrs.set(attr.getName(), attr.getValue());
+  /// Note: \p AddAttrPtr is used to provide a concrete implementation
+  /// controlling where to add the attribute (to the 'passthrough' list or not).
+  AttrBuilder &
+  addAttribute(llvm::Twine AttrName, mlir::Attribute Attr,
+               AddAttrFuncPtr AddAttrPtr = &AttrBuilder::addAttribute);
+
+  /// Add the given named attribute \p Attr to the builder.
+  AttrBuilder &addAttribute(mlir::NamedAttribute Attr) {
+    Attrs.set(Attr.getName(), Attr.getValue());
     return *this;
   }
 
-  /// Add the LLVM attribute identified by \p kind to the builder "passthrough"
+  /// Add the LLVM attribute identified by \p Kind to the builder "passthrough"
   /// named attribute.
-  AttrBuilder &addPassThroughAttribute(llvm::Attribute::AttrKind kind);
+  AttrBuilder &addPassThroughAttribute(llvm::Attribute::AttrKind Kind);
+
+  /// Add the LLVM attribute identified by \p Kind with a type given by \p Ty
+  /// to the builder "passthrough" named attribute.
+  AttrBuilder &addPassThroughAttribute(llvm::Attribute::AttrKind Kind,
+                                       mlir::Type Ty);
+
+  /// Add the LLVM attribute identified by \p Kind with a value given by \p Val
+  /// to the builder "passthrough" named attribute.
+  AttrBuilder &addPassThroughAttribute(llvm::Attribute::AttrKind Kind,
+                                       uint64_t Val);
+
+  /// Create a NamedAttribute with name \p AttrName and value \p Attr and add it
+  /// to the builder "passthrough" named attribute.
+  AttrBuilder &addPassThroughAttribute(llvm::StringRef AttrName,
+                                       mlir::Attribute Attr);
 
-  /// Add the given attribute \p attr to the builder "passthrough" named
+  /// Add the given named attribute \p Attr to the builder "passthrough" named
   /// attribute.
-  AttrBuilder &addPassThroughAttribute(mlir::Attribute attr);
+  AttrBuilder &addPassThroughAttribute(mlir::NamedAttribute Attr);
+
+  /// Add \p Attr to the \p PassThroughAttr list.
+  /// Note: \p PassThroughAttr must have name "passthrough" and value of type
+  /// ArrayAttr.
+  static void addToPassThroughAttr(mlir::NamedAttribute &PassThroughAttr,
+                                   mlir::NamedAttribute Attr,
+                                   mlir::MLIRContext &Ctx);
+
+  /// Add \p NewAttrs into the attribute list of \p PassThroughAttr.
+  /// Note: \p PassThroughAttr must have name "passthrough" and value of type
+  /// ArrayAttr.
+  static void addToPassThroughAttr(mlir::NamedAttribute &PassThroughAttr,
+                                   mlir::ArrayAttr NewAttrs,
+                                   mlir::MLIRContext &Ctx);
```

Comment:
No need to add this to the interface, please keep in `.cc`.

Reply:
Will move.